### PR TITLE
HDDS-11157. Improve Datanodes page UI

### PR DESCRIPTION
**Pull request by devabhishekpal**
## What changes were proposed in this pull request?
Improve Datanodes page UI.

Please describe your PR in detail:
* This PR adds a new Datanodes Page with improved layout and design.
* The new page would help to eventually improve the overall look and feel for Recon
* Currently the page is disabled to the user so as to not affect any existing functionality

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11157

## How was this patch tested?
Patch was tested manually
Screenshot 2024-09-06 at 00 26 13
Screenshot 2024-09-06 at 00 26 30
Screenshot 2024-09-06 at 00 34 24
Screenshot 2024-09-06 at 00 40 21
Screenshot 2024-09-06 at 00 34 34
Screenshot 2024-09-06 at 00 26 02

**Important**
There are some bugs in current version of AntD with Popover component which causes the Popover to display in incorrect position if we scroll the table.
Currently I am waiting for AntD dependencies to update in a new release so that previous license issues with the AntD slick component do not cause incompatibility.
After this I will target upgrade to AntD v5 which should take care of this bug.

### CODE REVIEW

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        <Descriptions.Item label="Decommissioning Started at">{decommissionStartTime}</Descriptions.Item>
+        <Descriptions.Item label="No. of Unclosed Pipelines">{numOfUnclosedPipelines}</Descriptions.Item>
+        <Descriptions.Item label="No. of Unclosed Containers">{numOfUnclosedContainers}</Descriptions.Item>
+        <Descriptions.Item label="No. of Under-Replicated Containers">{numOfUnderReplicatedContainers}</Descriptions.Item>
+        <Descriptions.Item label="Under-Replicated">{UnderReplicated}</Descriptions.Item>
+        <Descriptions.Item label="Unclosed">{UnClosed}</Descriptions.Item>
+      </Descriptions>
+    );
+  }
+  // Need to check summarydata is not empty
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx:105]: This comment seems not in sync what code is doing actually. Can we make it more meaningful ?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+  const { summaryData } = state;
+  if (summaryData?.datanodeDetails && summaryData?.metrics && summaryData?.containers) {
+    const {
+      datanodeDetails: { uuid, networkLocation, ipAddress, hostName },
+      containers: { UnderReplicated, UnClosed },
+      metrics: { decommissionStartTime, numOfUnclosedPipelines, numOfUnclosedContainers, numOfUnderReplicatedContainers }
+    } = summaryData as SummaryData;
+    content = (
+      <Descriptions size="small" bordered column={1} title={`Decommission Status: DECOMMISSIONING`}>
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx:93]: Can we move this whole `Description`  into its own function or component or clarity. Make it clumsy or less readable when nested inside main `if `block

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+const DecommissionSummary: React.FC<DecommisioningSummaryProps> = ({
+  uuid = ''
+}) => {
+  const [state, setState] = React.useState<DecommisioningSummaryState>({
+    summaryData: {},
+    loading: false
+  });
+  const cancelSignal = React.useRef<AbortController>();
+
+  async function fetchDecommissionSummary(selectedUuid: string) {
```
Comment (devmadhuu) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx:44]: Not sure if it is feasible, but check if we can add a loading spinner or skeleton to the UI to indicate that data is being fetched:

**Review by devmadhuu — COMMENTED** [review]
Thanks @devabhishekpal for working on this patch. Just few nits.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx (hunk 2 above)
Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx]: Addressed in the latest commit [d9d12df](https://github.com/apache/ozone/pull/7168/commits/d9d12dfe4ef4378753d36c1e58f1c6238b85fdbd)

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx (hunk 1 above)
Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/decommissioningSummary/decommissioningSummary.tsx]: Sorry, this was a previous comment which is not applicable after cleanup, removed in latest commit

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+          trigger="hover">
+          <strong>{pipelines.length}</strong> pipelines
+        </Popover>
+      );
+    }
+  },
+  {
+    title: () => (
+      <span style={headerIconStyles} >
+        Leader Count
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:183]: In the old UI, we had a searchable Leader Count column, but in the new UI, this search functionality is missing. The same applies to the Container ID column.

image

Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:183]: I don’t think adding a search feature for Leader Count is necessary, but I do believe having a search option for Container ID is important. While debugging, we need to be able to identify which node contains a problematic container. It turns out that the search feature for the Container ID column was faulty and didn’t work correctly in the old UI.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+          <InfoCircleOutlined style={{ paddingLeft: '4px' }} />
+        </Tooltip>
+      </span>
+    ),
+    dataIndex: 'openContainers',
+    key: 'openContainers',
+    sorter: (a: Datanode, b: Datanode) => a.openContainers - b.openContainers
+  },
+  {
+    title: 'Version',
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:214]: We will also need to add search functionality for the Version column in the new UI.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    title: 'Setup Time',
+    dataIndex: 'setupTime',
+    key: 'setupTime',
+    sorter: (a: Datanode, b: Datanode) => a.setupTime - b.setupTime,
+    render: (uptime: number) => {
+      return uptime > 0 ? moment(uptime).format('ll LTS') : 'NA';
+    }
+  },
+  {
+    title: 'Revision',
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:230]: I’m not sure if the searchable Revision field is useful in its current form.

image

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+];
+
+const DatanodesTable: React.FC<DatanodeTableProps> = ({
+  data,
+  handleSelectionChange,
+  decommissionUuids,
+  selectedColumns,
+  loading = false,
+  selectedRows = [],
+  searchColumn = 'hostname',
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:259]: This issue is present in both the old and new UI: if a user accidentally adds a space at the end of their search input, the search doesn’t work. While it’s not a critical issue, we could remove any unnecessary spaces from the search input automatically. It’s a small improvement, but I noticed it when I was testing the search functionality.

This issue is present in both the old and new UI: if a user accidentally adds a space at the end of their search input, the search doesn’t work. While it’s not a critical issue, we could remove any unnecessary spaces from the search input automatically. It’s a small improvement, but I noticed it when I was testing the search functionality.

This issue is present in both the old and new UI: if a user accidentally adds a space at the end of their search input, the search doesn’t work. While it’s not a critical issue, we could remove any unnecessary spaces from the search input automatically. It’s a small improvement, but I noticed it when I was testing the search functionality.

image

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/pages/datanodes/datanodes.tsx [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+          width={400} >
+            <div style={{
+              margin: '0px 0px 5px 0px',
+              fontSize: '16px',
+              fontWeight: 'bold'
+            }}>
+              <WarningFilled className='icon-warning' style={{paddingRight: '8px'}}/>
+              Stop Tracking Datanode
+            </div>
+            Are you sure, you want recon to stop tracking the selected <strong>{selectedRows.length}</strong> datanode(s)?
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/pages/datanodes/datanodes.tsx:306]: Another issue I encountered is that when all datanodes are removed from the list (e.g., by shutting them down), the new UI crashes and displays an error. This behavior wasn’t present in the old UI, which handled this case by showing an empty list.

image

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx (hunk 6 above)
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:230]: Let’s discuss offline whether these additional search functionalities are required for certain columns, or if we should limit it to only the use cases where it is really necessary and will assist users in debugging.

**Review by ArafatKhan2198 — CHANGES_REQUESTED** [review]
Thank you for your tremendous effort on this improvement, @devabhishekpal. I've left some comments for your review.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx (hunk 4 above)
Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:183]: I believe we can go over the searchable fields. For leader count it was just a number and I didn't think it was useful making it searchable

Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/components/tables/datanodesTable.tsx:183]: So this was the containers count column and as discussed this won't benefit from a search feature.

**Comment by ArafatKhan2198** [comment]
I noticed Devesh's comment about organizing different functionalities into separate files. Do we need something similar for this page as well, @devabhishekpal?
https://github.com/apache/ozone/pull/7171#issuecomment-2342794750

**Comment by devabhishekpal** [comment]
> I noticed Devesh's comment about organizing different functionalities into separate files. Do we need something similar for this page as well, @devabhishekpal? [#7171 (comment)](https://github.com/apache/ozone/pull/7171#issuecomment-2342794750)

So I have separated out the table into it's own component. However for the others please refer to https://github.com/apache/ozone/pull/7171#issuecomment-2341729481 for reason

**Review by ArafatKhan2198 — APPROVED** [review]
Thanks for updating the patch @devabhishekpal 
LGTM +1

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/pages/datanodes/datanodes.tsx (hunk 8 above)
Comment (devmadhuu) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/pages/datanodes/datanodes.tsx:303]: Is this issue taken care ?

Comment (devabhishekpal) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/v2/pages/datanodes/datanodes.tsx:303]: Yes @devmadhuu, this has been fixed and tested out after building the new changes as well

**Review by devmadhuu — COMMENTED** [review]
Thanks @devabhishekpal for addressing the comments. Few behaviour changes, pls check:
1. For Datanode search, adding space in search text, still issue not resolved.
2. In Old UI, we had search for `State` and `Operational State` which is missing from new UI.

**Review by devmadhuu — APPROVED** [review]
Thanks @devabhishekpal  for addressing comments and clarifying with detailed reason. Patch LGTM +1

**Comment by ArafatKhan2198** [comment]
Thanks @devabhishekpal for the patch & @devmadhuu for the review.

**Pull request merged**